File: 2020/24_LobbyLayout/tiles.lua
-- ======================================================================
-- Lobby Layout
--   Advent of Code 2020 Day 24 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         t i l e s . l u a
-- ======================================================================
-- A solver for the Advent of Code 2020 Day 24 puzzle

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Tiles = { part2 = false, text = {}, tiles = {} }

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------
local NUMBER_OF_DAYS = 100

local DELTA = {
  E = {10, 0},
  W = {-10, 0},
  NE = {5, -5},
  SE = {5, 5},
  NW = {-5, -5},
  SW = {-5, 5}
}

local BLACK = true
local WHITE = false

local ROWMULT = 100000
-- ======================================================================
--                                                                  Tiles
-- ======================================================================

-- Object for Lobby Layout

function Tiles:Tiles (o)

  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or {}
  o.tiles = {}

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:process_text(o.text)
  end

  return o
end

function Tiles:process_text(text)
  -- Assign values from text

  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)

  -- 1. Loop for each line of the text
  for _, line in ipairs(text) do

    -- 2. Determine the location of this tile
    local loc = self:location(line)
    -- print(loc, line)
    -- 3. Set (or reset) the tile
    if nil == self.tiles[loc] then
      self.tiles[loc] = BLACK
    else
      self.tiles[loc] = nil
    end
  end
end

function Tiles:location(line)
  -- Determine the location of the tile by following the direction instructions
  
  -- 1. Isolate the instructions
  local instructions = string.gsub(line, "nw", "NW,")
  instructions = string.gsub(instructions, "ne", "NE,")  
  instructions = string.gsub(instructions, "sw", "SW,")  
  instructions = string.gsub(instructions, "se", "SE,")  
  instructions = string.gsub(instructions, "w", "W,")  
  instructions = string.gsub(instructions, "e", "E,")  
  
  -- 2. Start at the reference tile
  local loc = {10000, 10000}
  
  -- 3. Loop for all of the instructions
  for inst in instructions:gmatch("%a+") do
    
    -- 4. Adjust the location
    loc[1] = loc[1] + DELTA[inst][1]
    loc[2] = loc[2] + DELTA[inst][2]
  end
  
  -- 5. Return the final location as a single number
  return loc[2] * ROWMULT + loc[1]
end
  
function Tiles:how_many_black()
  -- Return count of black tiles
  
  -- 1. Start with none
  local result = 0
  
  -- 2. Loop through all the tiles
  for _, side in pairs(self.tiles) do
    -- print(loc, side)
    
    -- 3. If the black side is up, increment count
    if side then
      result = result + 1
    end
  end
  
  -- 4. Return the number of tiles with the black side showing
  return result
end

function Tiles:tile(row, col)
  -- Return the status (BLACK or WHITE) of a tile
  
  -- 1. Convert the row and column to an index
  local index = row * ROWMULT + col 
  
  -- 2. If no tile saved, then it is white
  if nil == self.tiles[index] then
    return WHITE
  end
  
  -- 3. Return the status of the tile
  return self.tiles[index]
end


function Tiles:neighbors(row, col)
  -- Return the tiles next to the specified tile
  
  -- 1. Start with nothing
  local result = {}
  
  -- 2. Loop for all of the neighbors
  for _, delta in pairs(DELTA) do
    
    -- 3. Get the location of the nearby tile
    local loc = (row + delta[2]) * ROWMULT + delta[1] + col
    
    -- 4. Add the locations
    table.insert(result, loc)
  end
  
  -- 5. Return the nearby tile locations
  return result
end
  
function Tiles:black_neighbors(nearby)
  -- Return the number of neighboring black tiles
  
  -- 0. Precondition axions
  assert(#nearby == 6)
  
  -- 1. Start with nothing
  local result = 0
  
  -- 2. Loop for all of the neighbors
  for _, loc in ipairs(nearby) do
    
    -- 3. Get the status of the nearby tile
    -- print(loc, self.tiles[loc])
    local side = self.tiles[loc]
    
    -- 4. If the tile is black, increment the count
    if side == BLACK then
      result = result + 1
    end
  end
  
  -- 5. Return the number of nearby black tiles
  return result
end

function Tiles:next_day()
  -- Advance tiles, returns true is something changed
  
  -- 1. Start with nothing
  local future = {}
  local changed = false
  local checked = {}
  
  -- 2. Loop for all of the current tiles (should all be black)
  for loc, side in pairs(self.tiles) do
    
    -- 3. Split the localtion into column and row
    local col = loc % ROWMULT
    local row = loc // ROWMULT
    -- print("next_day", loc, side, row, col)
      
    -- 3. Get the locations of nearby tiles
    local black_neighbors = self:neighbors(row, col)
    
    -- 4. Get the count of nearby black tiles
    local black_count = self:black_neighbors(black_neighbors)
            
    -- 5. we keep black tiles if it has exactly one or two neighbors
    if side == BLACK then
      if black_count == 1 or black_count == 2 then
        future[loc] = BLACK
        -- print("keeping black", loc)
      else  
        -- print("black to white", loc)
        changed = true
      end
      
      -- 6. Check the white tiles near the black one
      for _, nearby in ipairs(black_neighbors) do
        if self.tiles[nearby] ~= BLACK and nil == checked[nearby] then
          
          local white_col = nearby % ROWMULT
          local white_row = nearby // ROWMULT
          local white_neighbors = self:neighbors(white_row, white_col)
          local white_count = self:black_neighbors(white_neighbors)
          
          -- 7. A while tile changes to black if there are exactly two black tiles nearby
          if white_count == 2 then
            -- print("Setting to black", nearby)
            future[nearby] = true
            changed = true
          end
        end
        checked[nearby] = true
      end
        
    else  
      -- 8. A while tile changes to black if there are exactly two black tiles nearby 
      if black_count == 2 then
        future[loc] = BLACK
        changed = true
      end
    end
  end
  
  -- 9. The future is now
  self.tiles = future
  
  -- 10. Return true if any tile changed
  return changed
end

function Tiles:multiple_days(days)
  -- Advance the floor many times
  
  -- 1. Loop for the number of days
  for _ = 1, days do
    
    -- 2. Advance one day
    self:next_day()
  end
  
  -- 3. Return the number of black files
  return self:how_many_black()
end

function Tiles:part_one(args)
  -- Returns the solution for part one

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part one
  return self:how_many_black()
end


function Tiles:part_two(args)
  -- Returns the solution for part two

  -- 0. Precondition axioms
  local verbose = args.verbose or false
  local limit = args.limit or 0
  assert(verbose == true or verbose == false)
  assert(limit >= 0)

  -- 1. Return the solution for part two
  return self:multiple_days(NUMBER_OF_DAYS)
end

-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------

return Tiles

-- ======================================================================
-- end                        t i l e s . l u a                       end
-- ======================================================================

File: 2020/24_LobbyLayout/tiles.ts
// ======================================================================
// Lobby Layout
//   Advent of Code 2020 Day 24 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           t i l e s . t s
//
// A solver for the Advent of Code 2020 Day 24 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                  types
// ----------------------------------------------------------------------
type Delta = Record<string, number[]>;
// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
// const NUMBER_OF_DAYS = 100;

const DELTA: Delta = {
  E: [10, 0],
  W: [-10, 0],
  NE: [5, -5],
  SE: [5, 5],
  NW: [-5, -5],
  SW: [-5, 5],
};
const DELTA_KEYS = ['E', 'W', 'NE', 'SE', 'NW', 'SW'];

const CENTER = 10000;
const ROWMULT = 100000;

const RE_NW = RegExp('nw', 'g');
const RE_NE = RegExp('ne', 'g');
const RE_SW = RegExp('sw', 'g');
const RE_SE = RegExp('se', 'g');
const RE_W = RegExp('w', 'g');
const RE_E = RegExp('e', 'g');

const NUMBER_OF_DAYS = 100;

// ======================================================================
//                                                                  Tiles
// ======================================================================

export class Tiles {
  // Object for Lobby Layout
  text: string[];

  part2: boolean;

  tiles: Set<number>;

  constructor(text: string[], part2 = false) {
    // Create a Tiles object

    // 1. Set the initial values
    this.text = text === undefined ? [] : text;
    this.part2 = part2 === undefined ? false : part2;
    this.tiles = new Set();

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      this.processText(this.text);
    }
  }

  processText(text: string[]) {
    // Assign values from text

    // 1. Loop for every line of the text
    for (let indx = 0; indx < text.length; indx += 1) {
      const line = text[indx];

      // 2. Determine the location of this tile
      const loc = this.location(line);

      // 3. Set (or reset) the tile
      if (this.tiles.has(loc)) {
        this.tiles.delete(loc);
      } else {
        this.tiles.add(loc);
      }
    }
  }

  // eslint-disable-next-line class-methods-use-this
  location(line: string): number {
    // Determine the location of the tile by following the direction instructions

    // 1. Isolate the instructions
    let instructions = line.replace(RE_NW, 'NW,');
    instructions = instructions.replace(RE_NE, 'NE,');
    instructions = instructions.replace(RE_SW, 'SW,');
    instructions = instructions.replace(RE_SE, 'SE,');
    instructions = instructions.replace(RE_W, 'W,');
    instructions = instructions.replace(RE_E, 'E,');
    const insts = instructions.split(',');

    // 2. Start at the reference tile
    const loc = [CENTER, CENTER];

    // 3. Loop for all of the instructions
    for (let indx = 0; indx < insts.length; indx += 1) {
      const inst = insts[indx];
      if (inst.length > 0) {
        // 4. Adjust the locations
        loc[0] += DELTA[inst][0];
        loc[1] += DELTA[inst][1];
      }
    }

    // 5. Return the final locations as a single number
    return loc[1] * ROWMULT + loc[0];
  }

  howManyBlack(): number {
    // Return count of black tiles
    return this.tiles.size;
  }

  isTileBlack(row: number, col: number): boolean {
    // Return true if the tile is black

    // 1. Covert the column and row number to an index
    const index = row * ROWMULT + col;

    // 2. If the tile is in the set of tiles then it is a black tile
    return this.tiles.has(index);
  }

  // eslint-disable-next-line class-methods-use-this
  neighbors(row: number, col: number): number[] {
    // Return the indexes of the tiles next to the specified tile

    // 1. Start with nothing
    const result = [];

    // 2. Loop for all of the neighbors
    for (let indx = 0; indx < DELTA_KEYS.length; indx += 1) {
      const delta = DELTA[DELTA_KEYS[indx]];

      // 3. Get the location of the neighborly tile
      const loc = (row + delta[1]) * ROWMULT + delta[0] + col;

      // 4. Add it to the result
      result.push(loc);
    }

    // 5. Return the locations of nearby files
    return result;
  }

  blackNeighbors(neighbors: number[]): number {
    // Return the number of black neighbots

    // 1. Start with nothing
    let result = 0;

    // 2. Loop for the neighbor indexes
    for (let indx = 0; indx < neighbors.length; indx += 1) {
      const neighbor = neighbors[indx];

      // 3. If the neighboring tile is black, increment the cound
      if (this.tiles.has(neighbor)) {
        result += 1;
      }
    }

    // 4. Return the number of nearby black tiles
    return result;
  }

  nextDay(): boolean {
    // Advance tiles, returns true if something changed

    // 1. Start with nothing
    const future: Set<number> = new Set();
    const checked: Set<number> = new Set();
    let changed = false;

    // 2. Loop for all of the current black tiles
    const current = Array.from(this.tiles.values());
    for (let indx = 0; indx < current.length; indx += 1) {
      const curindx = +current[indx];

      // 3. Split out the row and column
      const col = curindx % ROWMULT;
      const row = Math.floor(curindx / ROWMULT);

      // 4. Get the number of black neighbors
      const neighbors = this.neighbors(row, col);
      const blackCount = this.blackNeighbors(neighbors);

      // 5. We keep a black tile if it has exactly one or two neighbors
      if (blackCount === 1 || blackCount === 2) {
        future.add(curindx);
      } else {
        changed = true;
      }

      // 6. Check the white tiles near the black one
      for (let nindx = 0; nindx < neighbors.length; nindx += 1) {
        const neighbor = +neighbors[nindx];
        if (!this.tiles.has(neighbor) && !checked.has(neighbor)) {
          const whiteCol = neighbor % ROWMULT;
          const whiteRow = Math.floor(neighbor / ROWMULT);
          const whiteNeighbors = this.neighbors(whiteRow, whiteCol);
          const whiteCount = this.blackNeighbors(whiteNeighbors);

          // 7. A white tile changes to black if there are exactly two black neighbors
          if (whiteCount === 2) {
            future.add(neighbor);
            changed = true;
          }

          // 8. We have checked this location for the future
          checked.add(neighbor);
        }
      }
    }

    // 9. The future is now
    this.tiles = future;

    // 10. Return true if any tile changed
    return changed;
  }

  multipleDays(days: number): number {
    // Advance several days and return the number of black tiles

    // 1. Loop for the number of days
    for (let day = 0; day < days; day += 1) {
      // 2. Advance one day
      this.nextDay();
    }
    // 3. Return the number of black tiles
    return this.howManyBlack();
  }

  solution(verbose = false, limit = 0): number {
    if (verbose) console.log(`solution: ${limit}`);
    if (this.part2) {
      return this.multipleDays(NUMBER_OF_DAYS);
    }
    return this.howManyBlack();
  }

  partOne(verbose = false, limit = 0): number {
    // Returns the solution for part one

    return this.solution(verbose, limit);
  }

  partTwo(verbose = false, limit = 0): number {
    // Returns the solution for part two

    // 1. Return the solution for part two
    return this.solution(verbose, limit);
  }
}

// ======================================================================
// end                         t i l e s . t s                        end
// ======================================================================
